ui_components/src/village/VillageUpgrades.js

```diff
@@ -0,0 +1,754 @@
+        apiFetch(
+            villageAPI,
+            {
+                request: 'GetConstructionBoostCost',
+                building_key: selectedBuilding.key,
+            }
+        ).then((response) => {
+            if (response.errors.length) {
+                handleErrors(response.errors);
+                return;
+            }
+            openModal({
+                header: 'Confirmation',
+                text: "Boosting the construction of this building will cost " + response.data.response_message + " Village Points",
+                ContentComponent: null,
+                onConfirm: () => BoostConstruction(),
+            });
+        });
+    }
+    const CheckBoostResearch = () => {
+        apiFetch(
+            villageAPI,
+            {
+                request: 'GetResearchBoostCost',
+                upgrade_key: selectedUpgrade.key,
+            }
+        ).then((response) => {
+            if (response.errors.length) {
+                handleErrors(response.errors);
+                return;
+            }
+            openModal({
+                header: 'Confirmation',
+                text: "Boosting the research of this upgrade will cost " + response.data.response_message + " Village Points",
+                ContentComponent: null,
+                onConfirm: () => BoostResearch(),
+            });
+        });
+    }
+    const BoostConstruction = () => {
+        apiFetch(
+            villageAPI,
+            {
+                request: 'CreateProposal',
+                type: 'boost_construction',
+                building_key: selectedBuilding.key,
+            }
+        ).then((response) => {
+            if (response.errors.length) {
+                handleErrors(response.errors);
+                return;
+            }
+            setProposalDataState(response.data.proposalData);
+            openModal({
+                header: 'Confirmation',
+                text: response.data.response_message,
+                ContentComponent: null,
+                onConfirm: null,
+            });
+        });
+    }
+    const BoostResearch = () => {
+        apiFetch(
+            villageAPI,
+            {
+                request: 'CreateProposal',
+                type: 'boost_research',
+                upgrade_key: selectedUpgrade.key,
+            }
+        ).then((response) => {
+            if (response.errors.length) {
+                handleErrors(response.errors);
+                return;
+            }
+            setProposalDataState(response.data.proposalData);
+            openModal({
+                header: 'Confirmation',
+                text: response.data.response_message,
+                ContentComponent: null,
+                onConfirm: null,
+            });
+        });
+    }
+    const buildingClickHandler = (building) => {
+        setSelectedUpgrade(null);
+        setSelectedBuilding(building);
+    };
+    const renderBuildingDetails = () => {
```

Comment:
Could this function be its own component, with the data it needs passed in explicitly?

Same comment generally applies anytime you have a `renderX` function it may be a sign to pull that into a separate component (same file OK)